File: fetch-chaos-icons.py
#!/usr/bin/env python3

"""
Fetch and update icons for Chaos Selfhosted programs
Adds Google favicon icons to all programs in chaos_selfhosted.json
"""

import json
from urllib.parse import urlparse


def update_chaos_icons():
    """Update icons for all programs in chaos_selfhosted.json"""
    
    program_file = "chaos_selfhosted.json"
    
    try:
        # Load programs
        with open(program_file, "r") as f:
            programs = json.load(f)
        
        print(f"[*] Loaded {len(programs)} programs from {program_file}")
        
        updated = 0
        
        # Add icons to each program
        for p in programs:
            try:
                if "program_url" in p and p["program_url"]:
                    domain = urlparse(p["program_url"]).netloc
                    if domain:
                        p["icon"] = f"https://www.google.com/s2/favicons?domain={domain}&sz=64"
                        updated += 1
            except Exception as e:
                print(f"[!] Error processing {p.get('name', 'Unknown')}: {e}")
                p["icon"] = ""
        
        # Save updated programs
        with open(program_file, "w") as f:
            json.dump(programs, f, indent=2)
        
        print(f"[✓] Icons added to {updated} programs")
        print("\n[*] Sample programs:")
        for p in programs[:5]:
            print(f"  {p['name']:30} | {p['icon']}")
    
    except FileNotFoundError:
        print(f"[✗] File not found: {program_file}")
    except json.JSONDecodeError:
        print(f"[✗] Invalid JSON in {program_file}")
    except Exception as e:
        print(f"[✗] Error: {e}")


if __name__ == "__main__":
    update_chaos_icons()

File: fetch-chaos-icons.js
#!/usr/bin/env node

/**
 * Fetch and update icons for Chaos Selfhosted programs
 * Adds Google favicon icons to all programs in chaos_selfhosted.json
 */

const fs = require('fs');
const https = require('https');
const { URL } = require('url');

const PROGRAM_FILE = 'chaos_selfhosted.json';

function extractDomain(url) {
    try {
        const parsed = new URL(url);
        return parsed.hostname;
    } catch {
        return '';
    }
}

function getFaviconUrl(domain) {
    return `https://www.google.com/s2/favicons?domain=${domain}&sz=64`;
}

async function updateChaosIcons() {
    try {
        // Read programs
        if (!fs.existsSync(PROGRAM_FILE)) {
            console.error(`[✗] File not found: ${PROGRAM_FILE}`);
            process.exit(1);
        }

        const programs = JSON.parse(fs.readFileSync(PROGRAM_FILE, 'utf-8'));
        console.log(`[*] Loaded ${programs.length} programs from ${PROGRAM_FILE}`);

        let updated = 0;

        // Add icons to each program
        for (const program of programs) {
            if (program.program_url) {
                const domain = extractDomain(program.program_url);
                if (domain) {
                    program.icon = getFaviconUrl(domain);
                    updated++;
                }
            }
        }

        // Save updated programs
        fs.writeFileSync(PROGRAM_FILE, JSON.stringify(programs, null, 2));

        console.log(`[✓] Icons added to ${updated} programs`);
        console.log('\n[*] Sample programs:');
        programs.slice(0, 5).forEach((p) => {
            console.log(`  ${p.name.padEnd(30)} | ${p.icon}`);
        });

    } catch (error) {
        console.error('[✗] Error:', error.message);
        process.exit(1);
    }
}

updateChaosIcons();